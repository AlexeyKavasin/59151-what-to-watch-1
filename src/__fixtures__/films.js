export default [
  {
    name: `Fantastic Beasts`,
    poster: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    id: 0,
    trailer: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    isPlaying: false,
    genre: `Comedy`
  },
  {
    name: `Bohemian Rhapsody`,
    poster: `img/bohemian-rhapsody.jpg`,
    id: 1,
    trailer: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    isPlaying: false,
    genre: `Comedy`
  },
  {
    name: `Macbeth`,
    poster: `img/macbeth.jpg`,
    id: 2,
    trailer: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    isPlaying: false,
    genre: `Comedy`
  },
  {
    name: `Aviator`,
    poster: `img/Aviator.jpg`,
    id: 3,
    trailer: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    isPlaying: false,
    genre: `Comedy`
  },
  {
    name: `Pulp Fiction`,
    poster: `img/pulp-fiction.jpg`,
    id: 4,
    trailer: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    isPlaying: false,
    genre: `Documentary`
  }
];
