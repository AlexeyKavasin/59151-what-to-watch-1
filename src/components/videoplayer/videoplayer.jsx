import React from "react";
import PropTypes from "prop-types";

export class VideoPlayer extends React.PureComponent {
  constructor(props) {
    super(props);

    this._videoRef = React.createRef();
    this.playerTimeout = null;
    this.playerDelay = 1000;
  }

  render() {
    const {trailer, poster} = this.props;

    return <React.Fragment>
      <video
        ref={this._videoRef}
        poster={poster}
        width="280"
        height="175"
        muted>
        <source src={trailer}></source>
      </video>
    </React.Fragment>;
  }

  componentDidUpdate(prevProps) {
    const video = this._videoRef.current;
    clearTimeout(this.playerTimeout);

    if (prevProps.isPlaying !== this.props.isPlaying) {
      if (this.props.isPlaying) {
        this.playerTimeout = setTimeout(() => {
          video.play();
        }, this.playerDelay);
      } else {
        video.pause();
        video.load();
      }
    }
  }
}

VideoPlayer.propTypes = {
  trailer: PropTypes.string,
  poster: PropTypes.string,
  isPlaying: PropTypes.bool.isRequired
};
