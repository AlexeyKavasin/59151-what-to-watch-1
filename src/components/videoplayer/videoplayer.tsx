import * as React from "react";
import {IVideoPlayer} from "../../interfaces";

export class VideoPlayer extends React.PureComponent<IVideoPlayer, null> {
  private _videoRef: React.RefObject<HTMLVideoElement>;
  private playerTimeout: null | NodeJS.Timer;
  private playerDelay: number;

  constructor(props) {
    super(props);

    this._videoRef = React.createRef();
    this.playerTimeout = null;
    this.playerDelay = 1000;
  }

  render() {
    const {trailer, poster} = this.props;

    return <React.Fragment>
      <video
        ref={this._videoRef}
        poster={poster}
        width="280"
        height="175"
        muted>
        <source src={trailer} />
      </video>
    </React.Fragment>;
  }

  componentDidUpdate(prevProps) {
    const video = this._videoRef.current;
    clearTimeout(this.playerTimeout);

    if (prevProps.isPlaying !== this.props.isPlaying) {
      if (this.props.isPlaying) {
        this.playerTimeout = setTimeout(() => {
          video.play();
        }, this.playerDelay);
      } else {
        video.pause();
        video.load();
      }
    }
  }
}
