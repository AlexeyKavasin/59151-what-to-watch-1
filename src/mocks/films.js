export default [
  {
    name: `Fantastic Beasts`,
    src: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`
  },
  {
    name: `Bohemian Rhapsody`,
    src: `img/bohemian-rhapsody.jpg`
  },
  {
    name: `Macbeth`,
    src: `img/macbeth.jpg`
  },
  {
    name: `Aviator`,
    src: `img/Aviator.jpg`
  },
  {
    name: `Pulp Fiction`,
    src: `img/pulp-fiction.jpg`
  }
];
