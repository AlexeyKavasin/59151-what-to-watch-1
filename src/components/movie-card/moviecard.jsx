import React from "react";
import {VideoPlayer} from "../videoplayer/videoplayer.jsx";
import {propTypes, defaultProps} from "../../proptypes/moviecard.props";

export class MovieCard extends React.PureComponent {
  constructor(props) {
    super(props);
    this.handlePlayClick = this.handlePlayClick.bind(this);
    this.handleMouseOver = this.handleMouseOver.bind(this);
  }

  handlePlayClick(event) {
    const {onPlayClick, id} = this.props;
    onPlayClick(event, id);
  }

  handleMouseOver(event) {
    const {onMouseOver, id} = this.props;
    onMouseOver(event, id);
  }

  render() {
    const {name, poster, trailer, id} = this.props;
    return <React.Fragment>
      <article className="small-movie-card catalog__movies-card" id={id} onMouseOver={this.handleMouseOver}>
        <button className="small-movie-card__play-btn" type="button" onClick={this.handlePlayClick}>Play</button>
        <div className="small-movie-card__image">
          <VideoPlayer trailer={trailer} poster={poster} onMouseOver={this.handleMouseOver} isPlaying={false}/>
        </div>
        <h3 className="small-movie-card__title">
          <a className="small-movie-card__link" href="movie-page.html">{name}</a>
        </h3>
      </article>
    </React.Fragment>;
  }
}

MovieCard.propTypes = propTypes;
MovieCard.defaultProps = defaultProps;
