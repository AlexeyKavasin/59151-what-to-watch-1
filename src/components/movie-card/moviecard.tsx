import * as React from "react";
import {VideoPlayer} from "../videoplayer/videoplayer";
import {IMovieCard} from "../../interfaces";

export class MovieCard extends React.PureComponent<IMovieCard, null> {
  constructor(props) {
    super(props);
    this.handleMouseOver = this.handleMouseOver.bind(this);
    this.handleMouseLeave = this.handleMouseLeave.bind(this);
  }

  handleMouseOver(event) {
    const {onMouseOver, id} = this.props;
    onMouseOver(event, id);
  }

  handleMouseLeave() {
    const {onMouseLeave} = this.props;
    onMouseLeave();
  }

  render() {
    const {name, poster, trailer, id, isPlaying} = this.props;
    return <React.Fragment>
      <article className="small-movie-card catalog__movies-card" data-id={id} onMouseOver={this.handleMouseOver} onMouseLeave={this.handleMouseLeave}>
        <div className="small-movie-card__image">
          <VideoPlayer trailer={trailer} poster={poster} isPlaying={isPlaying}/>
        </div>
        <h3 className="small-movie-card__title">
          <a className="small-movie-card__link" href="movie-page.html">{name}</a>
        </h3>
      </article>
    </React.Fragment>;
  }
}
