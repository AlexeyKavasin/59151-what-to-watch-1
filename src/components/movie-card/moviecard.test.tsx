import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {MovieCard} from './moviecard';
import {VideoPlayer} from '../videoplayer/videoplayer';
import films from '../../__fixtures__/films.js';

it(`Movie card renders correctly`, () => {
  const film = films[0];
  const tree = renderer
    .create(<MovieCard {...film} isPlaying={false}>
      <VideoPlayer {...film} isPlaying={false}/>
    </MovieCard>, {createNodeMock: () => ({addEventListener() {}, removeEventListener() {}})})
    .toJSON();

  expect(tree).toMatchSnapshot();
});
