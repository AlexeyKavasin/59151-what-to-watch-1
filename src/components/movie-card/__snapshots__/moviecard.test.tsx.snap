// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie card renders correctly 1`] = `
<article
  className="small-movie-card catalog__movies-card"
  data-id={0}
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <a
    className="small-movie-card__image"
    href="#/film/0"
    onClick={[Function]}
  >
    <video
      height="175"
      muted={true}
      poster="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
      width="280"
    >
      <source
        src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
      />
    </video>
  </a>
  <h3
    className="small-movie-card__title"
  >
    <a
      className="small-movie-card__link"
      href="#/film/0"
      onClick={[Function]}
    >
      Fantastic Beasts
    </a>
  </h3>
</article>
`;
